core/src/main/java/org/apache/calcite/plan/Digest.java

```diff
@@ -0,0 +1,269 @@
+   * {@link RelNode} with a simple string format digest. **/
+  private static List<Pair<String, Object>> normalizeContents(
+      List<Pair<String, Object>> items) {
+    List<Pair<String, Object>> normalized = new ArrayList<>();
+    for (Pair<String, Object> item : items) {
+      if (item.right instanceof RelNode) {
+        RelNode input = (RelNode) item.right;
+        normalized.add(Pair.of(item.left, simpleRelDigest(input)));
+      } else {
+        normalized.add(item);
+      }
+    }
+    return normalized;
+  }
+
+  /**
+   * Returns a simple string format digest.
+   *
+   * <p>There are three kinds of nodes we need to handle:
+   *
+   * <ul>
+   * <li>RelSubset: composition of class name, set id and traits;</li>
+   * <li>HepRelVertex: composition of class name and current rel id;</li>
+   * <li>Normal rel: composition of class name and id.</li>
+   * </ul>
+   *
+   * @param rel The rel
+   */
+  private static String simpleRelDigest(RelNode rel) {
+    StringBuilder digest = new StringBuilder(rel.getRelTypeName());
+    digest.append('#');
+    if (rel instanceof RelSubset) {
+      RelSubset subset = (RelSubset) rel;
+      digest.append(subset.getSetId());
+      for (RelTrait trait : subset.getTraitSet()) {
+        digest.append('.').append(trait);
+      }
+    } else if (rel instanceof HepRelVertex) {
+      digest.append(((HepRelVertex) rel).getCurrentRel().getId());
+    } else {
+      digest.append(rel.getId());
+    }
+    return digest.toString();
+  }
+
+  @Override public String toString() {
+    if (null != digest) {
+      return digest;
+    }
+    StringBuilder sb = new StringBuilder();
+    sb.append(rel.getRelTypeName());
+
+    for (RelTrait trait : rel.getTraitSet()) {
+      sb.append('.');
+      sb.append(trait.toString());
+    }
+
+    sb.append('(');
+    int j = 0;
+    for (Pair<String, Object> item : items) {
+      if (j++ > 0) {
+        sb.append(',');
+      }
+      sb.append(item.left);
+      sb.append('=');
+      sb.append(item.right);
+    }
+    sb.append(')');
+    digest = sb.toString();
+    return digest;
+  }
+
+  @Override public int compareTo(Digest other) {
+    return this.rel.getId() - other.rel.getId();
+  }
+
+  @Override public boolean equals(Object o) {
+    if (this == o) {
+      return true;
+    }
+    if (o == null || getClass() != o.getClass()) {
+      return false;
+    }
+    Digest that = (Digest) o;
+    return hashCode == that.hashCode && deepEquals(that);
+  }
+
+  /**
+   * The method is used to resolve hash conflict, in current 6000+ tests, there are about 8
+   * tests with conflict, so we do not cache the hash code items in order to
+   * reduce mem consumption.
+   */
+  private boolean deepEquals(Digest other) {
+    Object[] thisItems = collect(this.rel, this.items, true);
+    Object[] thatItems = collect(other.rel, other.items, true);
+    if (thisItems.length != thatItems.length) {
```

Comment:
The whole check + loop could be replaced with `Arrays.equals(Object[], Object[])`

Reply:
Sorry, i would choose current code because it's simpler.

Reply:
are you telling that 9 lines are simpler that one line of code which is `return Arrays.equals(thisItems, thatItems)`?

Reply:
Did you see the code in `Arrays.equals(thisItems, thatItems)` ? Here the `thisItems` and `thatItems` are both never null.

Reply:
Considering null checks are often optimized out by the JVM, I'm not sure that the possible extra optimization is worth the extra code you wrote...

Reply:
It is often but not always, if you are not sure also, let's keep it as it is, which is also straight-forward code.

Reply:
Sorry, I am at a loss for arguments. The original reply was that the code was simpler than `Arrays.equals(Object[], Object[])`, a standard JDK method, whereas in fact, it is basically the same code, except for the check for null, which is often suppressed by the JVM compiler (cf https://shipilev.net/jvm/anatomy-quarks/25-implicit-null-checks/) and even so has minimal impact compared to the two calls to the `#collect` method where a list is created and converted later into an array...

Reply:
The test here https://github.com/apache/calcite/pull/2034 shows that there is negligible performance affect with `ArrayList.toArray`, so still, i would not follow you.